### ISSUE TITLE
Sanitize JobInternal

### ISSUE THREAD

**Issue by lpellegr** [ow2-proactive/scheduling#2306]
JobInternal implementation is really weird. 

Several changes should be made to improve readability and performance:

  - Replace HashMap implementation for tracking tasks by a [TreeMap](http://docs.oracle.com/javase/7/docs/api/java/util/TreeMap.html) that order entries by keys. Order seems already maintained by returning task id as hashcode. However, tasks are sometimes reordered and reinjected in the map (e.g. `InternalJob#prepareTasks`). Besides, it makes some assumptions about how HashMap is implemented (open addressing vs separate chaining)
  - Check if `prepareTasks` method is still required once TreeMap is used
  - Find where recreation and reinjection of task ids is made (e.g. in `SchedulerDBManager#newJobSubmitted`)
      - Check if these pieces of code are still required
  - Check if it is possible to merge all different accessors for the internal hash map (e.g. `getHMTasks`, `getITasks`, `getIHMTasks`, `getTasks`, etc.). Ideally, a shallow copy of the collection should always be returned. If not possible, at least a better naming should be used.
  - Check that accessors like `getHMTasks` (which returns a shallow copy) is not called inside a for loop multiple times when it is not required (like it was the case in `SchedulerDBManager#newJobSubmitted).

**Comment by marcocast** [comment]
Instead of TreeMap, we may consider as well HashTreePMap from [pCollection lib](http://pcollections.org)  as it provides immutability and better performance together with clean code. http://pcollections.googlecode.com/svn/trunk/docs/org/pcollections/HashTreePMap.html

**Comment by lpellegr** [comment]
Interesting. [ImmutableSortedMap](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/ImmutableSortedMap.html) from Guava should be also considered if it is possible to use an immutable map. This would avoid importing yet another library.

**Comment by fviale** [comment]
If [ImmutableSortedMap](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/ImmutableSortedMap.html) refers to an immutable map, it would not be well suited to the tasks maps as tasks are added dynamically during iterations/replications. Unless I missed something.

**Comment by lpellegr** [comment]
@fviale yes, that's why I said *if it is possible to use an immutable map*. It was a suggestion related to what Marco proposed.

**Comment by fviale** [comment]
We should as well consider LinkedHashMap, which preserves insertion order. As TaskIds are always incrementing, insertion order should match TaskIds order.
